custom_metrics/javascript.js

```diff
@@ -0,0 +1,81 @@
+return JSON.stringify({
+    'num-ajax-requests': (() => {
+        // Returns the number of ajax requests per page.
+        var ajax_apis = ['xmlhttprequest', 'fetch', 'beacon'];
+        return window.performance.getEntriesByType('resource').filter(r => ajax_apis.includes(r.initiatorType)).length;
+
+    })(),
+    // Returns the number of ajax requests using beacon.
+    'beacon_ajax_usage': window.performance.getEntriesByType('resource').filter(r => r.initiatorType === 'beacon').length,
+    // Returns the number of iframes per page.
+    'iframe': document.getElementsByTagName('iframe').length,
+    'requests_protocol': (() => {
+        // Returns the percentage of http 2 protocol used.
+        try {
+            var requests = $WPT_REQUESTS;
+            if (requests.length > 0) {
+                var ajaxs = window.performance.getEntriesByType('resource').filter(r => ['xmlhttprequest', 'fetch', 'beacon'].includes(r.initiatorType))
+                var resources = window.performance.getEntriesByType('resource').filter(r => r.initiatorType !== 'xmlhttprequest' && r.initiatorType !== 'fetch' && r.initiatorType !== 'beacon')
+                return JSON.stringify({
+                    'total_requests': requests.length,
+                    'total_h1': requests.filter(r => ["http/1.1", "http/1"].includes(r.protocol)).length,
+                    'ajax_h1': ajaxs.filter(r => ["http/1.1", "http/1"].includes(r.nextHopProtocol)).length,
+                    'resources_h1': resources.filter(r => ["http/1.1", "http/1"].includes(r.nextHopProtocol)).length,
+                    'total_h2': requests.filter(r => ["h2", "http/2", "http2"].includes(r.protocol)).length,
+                    'ajax_h2': ajaxs.filter(r => ["h2", "http/2", "http2"].includes(r.nextHopProtocol)).length,
+                    'resources_h2': resources.filter(r => ["h2", "http/2", "http2"].includes(r.nextHopProtocol)).length,
+                    'total_h3': requests.filter(r => r.protocol === "http/3").length,
+                    'ajax_h3': ajaxs.filter(e => r.nextHopProtocol === "http/3").length,
+                    'resources_h3': resources.filter(e => r.nextHopProtocol === "http/3").length,
+                });
+            }
+            else {
+                return null;
+            }
+        } catch (e) {
+            return null;
+        }
+    })(),
+    'num_webcomp_shadow': (() => {
+        // Returns the number of web components or shadow roots on a web page.
+        try {
+            var elements_with_hyphen = Array.from(document.body.getElementsByTagName("*")).filter(e => e.nodeName.includes('-'));
```

Comment:
Do you also need to enforce that the hyphen is in the middle of the node name, or can it be at the start/end?

Reply:
Good point, just curious do you know any framework or libraries that do it? So I can handle work on handling this usecase.

Reply:
covered already in `has_shadow_root.js`
Done in [f6a6e24](https://github.com/HTTPArchive/legacy.httparchive.org/pull/213/commits/f6a6e24231692b19551b0e317620088b3fafd4ee)